docs/datamodeling.rst

```diff
@@ -0,0 +1,330 @@
+'snake_case'.
+
+Using field setters
+-------------------
+
+You can configure whether or not the binding will attempt to use field setters
+if present with the ``useFieldSetters`` configuration option. When this is 
+``true``, the binding will attempt to find methods in your class that meet
+the following requirements:
+
+* Prefixed with ``set`` followed by the field name in ``PascalCase``;
+
+* Contain one parameter with the same type of the field;
+
+* The method is public and non-static;
+
+For example, creating a bean that represents the ``Person`` schema type:
+
+.. tabs::
+
+    .. code-tab:: java
+        :caption: Java
+
+        @EdgeDBType
+        public class Person {
+            private String name;
+            private int age;
+
+            public void setName(String name) {
+                this.name = name;
+            }
+
+            public void setAge(int age) {
+                this.age = age;
+            }
+
+            public String getName() {
+                return this.name;
+            }
+
+            public int getAge() {
+                return this.age;
+            }
+        }
+
+    .. code-tab:: sdl
+        :caption: Schema
+
+        module default {
+            type Person {
+                property name -> str;
+                property age -> int32;
+            }
+        }
+
+The driver will give priority to ``setName`` and ``setAge`` rather than 
+reflection to set the field values.
+
+Multi-link properties
+---------------------
+
+The JVM doesn't retain generic information for collection generics. To get 
+around this, you must specify the type of the collection with the 
+``@EdgeDBLinkType`` annotation.
+
+.. tabs::
+
+    .. code-tab:: java
+        :caption: Java
+
+        @EdgeDBType
+        public class Person {
+            public String name;
+            public int age;
+
+            @EdgeDBLinkType(Person.class)
+            public List<Person> friends;
+        }
+
+    .. code-tab:: sdl
+        :caption: Schema
+
+        module default {
+            type Person {
+                property name -> str;
+                property age -> int32;
+                multi link friends -> Person;
+            }
+        }
+
+The binding accepts any collection type that is: an array, a ``List<?>``, 
+assignable from a ``List<?>``, or a ``HashSet<?>``.
+
+Custom deserializers
+--------------------
+
+You can specify a constructor as a target for deserialization with the
+``@EdgeDBDeserializer`` annotation. A deserializer has 2 valid modes of 
+operation: enumeration consumers, and value consumers.
+
+Enumerator consumer
+^^^^^^^^^^^^^^^^^^^
+
+An enumerator consumer takes only one parameter, an ``ObjectEnumerator``, 
+which provides a direct handle to the deserialization pipeline. Calling
+``next()`` preforms the deserialization step for one element and returns
+an ``ObjectEnumerator.ObjectElement``, containing the name, type, and 
+value.
```

Comment:
I think the build system will not handle the commas after the inline code properly. (You may want to build to confirm. I wasn't able to review this built. More on that later in the overall review comment…)

We may want to come up with some other way to refer to these just to skirt around that. Maybe `an ``ObjectEnumerator`` object`? I'll leave the final solution to you since I'm not exactly sure what sort of entity `ObjectEnumerator` and then `ObjectEnumerator.ObjectElement` refer to.